_posts/2023-06-27-redis.md

```diff
@@ -0,0 +1,23 @@
+ 
+These facts only holds when the size of your payload and the number of connections remain relatively small. this easily jumps out the window with ever increasing load parameters. The threshold is unfortunately rather low at a high number connections and increased payload sizes. Modern large scale micro-services will easily have over 100 running instance at medium scale. And since most instances employing some pooling mechanism so as not to pay a connection cost for each request, A single redis instance is going to do a bit of work in maintaining those connections not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as [KeyDb](https://docs.keydb.dev), Snapchats drop-in replacement as an alternative to redis, employs multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at twitter. [Twoemproxy](https://github.com/twitter/twemproxy), or nutcracker as it is informally know, is in itself single threaded and employs key hashing to store keys in shards of multiple redis instance give you proper multiplexing. While it may look like this is susceptible to the original problems of a single threaded application, it is not necessarily the reality, because twoemproxy employs a single thread for each redis instance, turning the whole system into a multi threaded systems. Of course this is still susceptible to hotkeys. These might look like ideal solutions, setting up new infrastructure as an intermediary service introduces a new failure point and this is neither trivial nor ideal. But when done, there is a lot of net positives. 
```

Comment:
```suggestion
These facts only hold when the size of your payload and the number of connections remain relatively small. This easily jumps out the window with ever-increasing load parameters. The threshold is, unfortunately, rather low at a high number of connections and increased payload sizes. Modern large-scale micro-services will easily have over 100 running instances at medium scale. And since most instances employ some pooling mechanism so as not to pay a connection cost for each request, a single Redis instance is going to do a bit of work in maintaining those connections, not to talk of serving requests as they come through. To improve performance at medium to high loads, some projects such as [KeyDb](https://docs.keydb.dev), Snapchat's drop-in replacement as an alternative to Redis, employ a multithreaded approach and a bit of magic to sustain some high workloads. This has been touted to provide 5x performance over Redis. Another solution which I have seen used is to employ a proxy that multiplexes over multiple redises. One such proxy is the twoemproxy developed at Twitter. [Twoemproxy](https://github.com/twitter/twemproxy), or nutcracker as it is informally known, is in itself single-threaded and employs key hashing to store keys in shards of multiple redis instances, giving you proper multiplexing. While it may look like this is susceptible to the original problems of a single-threaded application, it is not necessarily the reality because twoemproxy employs a single thread for each Redis instance, turning the whole system into a multi-threaded system. Of course, this is still susceptible to hotkeys. These might look like ideal solutions, but setting up new infrastructure as an intermediary service introduces a new failure point, which is neither trivial nor ideal. But when done, there are a lot of net positives. 
```